205 imap hi l1b annotated direct events

tmplummer opened the pull request:
# Change Summary
Rework of Hi L1A DE processing 
Add  Hi L1B DE processing
Add Hi L1B processing to cli

## Overview
When I started, I was only planning on adding Hi L1B DE processing. I found some variables missing from the L1A DE product that had been added after the initial implementation so I had to update that code as well. Sorry for the large set of changes.

closes #205 

## New Dependencies

## New Files
- imap_processing/cdf/config/imap_hi_global_cdf_attrs.yaml
   - hi global attributes yaml file
- imap_processing/cdf/config/imap_hi_variable_attrs.yaml
   - hi variable attributes yaml file
- imap_processing/hi/l1b/hi_l1b.py
   - add l1b processing of hi de product
- imap_processing/tests/hi/conftest.py
   - move fixture into conftest.py for reuse

## Deleted Files
<!--List each deleted file and add one or more bullets with the rationale for why the file is being deleted-->
- deleted file 1
   - explanation for why file was deleted

## Updated Files
- imap_processing/hi/l1a/hi_l1a.py
   - Allow Path as input to hi_l1a function
- imap_processing/hi/l1a/science_direct_event.py
   - Add some missing variables to hi_l1a_de products
   - Integrate yaml based attribute manager
- imap_processing/hi/utils.py
   - add `sensor` attribute to HIAPID class
- imap_processing/tests/hi/test_hi_l1a.py
   - update test coverage for l1a de testing
- imap_processing/tests/hi/test_hi_l1b.py
   - add l1b de test coverage
- imap_processing/tests/hi/test_utils.py
   - add test coverage for HIAPID class

## Testing
2 unit tests added

tmplummer committed:
begin l1b modification for DE processing

tmplummer committed:
Add Hi cdf attribute yaml files

tmplummer committed:
Allow for Path input to hi_l1a
Add sensor attribute to HIAPID class
Add test coverage for HIAPID class

tmplummer committed:
Move test fixture for hi_l1a into conftest.py for reuse

tmplummer committed:
modify hi.l1a.science_direct_event to use CdfAttributeManager for global attributes
Add meta_event_met variable to hi L1A de product
Inject sensor name into L1A DE logical source global attribute

tmplummer committed:
Add hi L1B processing of direct event product
Add File_naming_convention global attribute schema file

tmplummer committed:
Move Hi L1A DE processing to use CdfAttributeManager
Update Hi L1A DE variable attributes
Add missing Hi L1A DE variables
Complete Hi L1B DE test

tmplummer committed:
Add hi l1b to hi processing in cli

tmplummer committed:
remove accidental addition of File_naming_convention to schema

tmplummer committed:
fix unit test

tech3371 commented on imap_processing/cdf/config/imap_hi_global_cdf_attrs.yaml:
```diff
@@ -0,0 +1,24 @@
+# YAML entries at the root level will overwrite the default global attributes
+Data_version: "001"
+Descriptor: &desc
+  Hi>IMAP High-Energy (IMAP-Hi) Energetic Neutral Atom Imager
+TEXT: &txt >
+  IMAP-Hi consists of two identical, single-pixel high-energy energetic
+  neutral atom (ENA) imagers mounted at fixed angles of 90 and 45 degrees
+  relative to the spacecraft spin axis. These imagers measure neutral atoms
+  entering our solar system from the outer edge of the heliosphere as they
+  move towards the Sun. See https://imap.princeton.edu/instruments/imap-hi for
+  more details.
+Instrument_type: Particles (space)
+
+imap_hi_l1a_de_attrs:
+    Data_level: 1A
+    Data_type: L1A_DE>Level-1A Direct Event
+    Logical_source: imap_hi_l1a_{sensor}-de
+    Logical_source_description: IMAP-Hi Instrument Level-1A Direct Event Data.
+
+imap_hi_l1b_de_attrs:
+    Data_level: 1B
```
```suggestion
# YAML entries at the root level will overwrite the default global attributes
instrument_base: &instrument_base
  Data_version: "001"
  Descriptor: &desc
    Hi>IMAP High-Energy (IMAP-Hi) Energetic Neutral Atom Imager
  TEXT: &txt >
    IMAP-Hi consists of two identical, single-pixel high-energy energetic
    neutral atom (ENA) imagers mounted at fixed angles of 90 and 45 degrees
    relative to the spacecraft spin axis. These imagers measure neutral atoms
    entering our solar system from the outer edge of the heliosphere as they
    move towards the Sun. See https://imap.princeton.edu/instruments/imap-hi for
    more details.
  Instrument_type: Particles (space)

imap_hi_l1a_de_attrs:
  <<: *instrument_base
  Data_level: 1A
  Data_type: L1A_DE>Level-1A Direct Event
  Logical_source: imap_hi_l1a_{sensor}-de
  Logical_source_description: IMAP-Hi Instrument Level-1A Direct Event Data.

imap_hi_l1b_de_attrs:
  <<: *instrument_base
  Data_level: 1B
```

tmplummer replied:
I started out with an implementation much like your suggestion and was having trouble figuring out why my `Data_version` string from this YAML was an integer when I retrieved it from the CDFManager. I found that the data manager code loops over each schema item and first checks if that key is in the global attributes before looking in the instrument specific global attributes. This meant that I was always getting the value back from the `Data_version` defined in the imap_default_global_cdf_attrs.yaml rather than the Hi specific `Data_version` defined in the `instrument_base`. By defining these values at the root level, when the CDFManager loads the file, the values overwrite the values in the global file which means I always get back the values that I want. To view the logic that I am talking about, see this function: https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/cdf/cdf_attribute_manager.py#L147

I didn't follow the bug that you and Laura found that caused the move to the `instrument_base` definition style, so I can't speak to that.

tech3371 replied:
@maxinelasp will be opening PR to handle this. You and her caught this bug in time.

tech3371 commented on imap_processing/cdf/config/imap_hi_variable_attrs.yaml:
```diff
@@ -0,0 +1,271 @@
+  VAR_TYPE: data
+  SCALE_TYPE: linear
+
+default_uint8_attrs: &default_uint8
```
cool. I really like these some other defaults option. Would love to do same for mine. In the future, this would be a nice shared defaults. Also I like that you are adding `dtype`. Would like to do same in the future.

tmplummer replied:
Yes! I agree that figuring out a way to add attribute inheritance across yaml files would be beneficial and allow everyone to use these type specific defaults.

tech3371 commented on imap_processing/cdf/config/imap_hi_variable_attrs.yaml:
```diff
@@ -0,0 +1,271 @@
+    Time of flight is 10-bit integer value that represents the time of flight of
+    the direct event. 1023 is the value used to indicate no event was registered.
+
+# ------- L1B DE Section -------
```
Comments like these helps

tech3371 commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -245,86 +252,60 @@ def create_dataset(de_data_list: list, packet_met_time: list) -> xr.Dataset:
+    cdf_manager.load_variable_attributes("imap_hi_variable_attrs.yaml")
+
+    # Inject sensor head into Logical_source
+    gattrs = cdf_manager.get_global_attributes("imap_hi_l1a_de_attrs")
```
can we spell out global?
```suggestion
    de_global_attrs = cdf_manager.get_global_attributes("imap_hi_l1a_de_attrs")
```

tech3371 commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -245,86 +252,60 @@ def create_dataset(de_data_list: list, packet_met_time: list) -> xr.Dataset:
-        dims="epoch",
-        attrs=hi_cdf_attrs.ccsds_met_attrs.output(),
-    )
+    for var_name, data in data_dict.items():
```
nice!

tech3371 commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -348,6 +329,7 @@ def science_direct_event(packets_data: list[Packet]) -> xr.Dataset:
     dataset: xarray.Dataset
         xarray dataset
     """
+    sensor_str = HIAPID(packets_data[0].header["PKT_APID"].raw_value).sensor
```
nice!

greglucas commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -178,21 +179,27 @@ def create_dataset(de_data_list: list, packet_met_time: list) -> xr.Dataset:
-    tof_3 = []
-    de_tag = []
-    ccsds_met = []
+    data_dict = {
```
As an FYI, no need to change here. This could be `data_dict = collection.defaultdict(list)` if you wanted to avoid explicitly writing all of the entries out, but it is also perhaps nice to explicitly list the items.

tmplummer replied:
Thanks for that. I hadn't dug into the documentation for what a `collection.defaultdict` was. This is useful.

tech3371 commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -42,9 +53,67 @@ def hi_l1b(l1a_cdf_path: Path):
+    #    some functionality can be found in imap_data_access.file_validation but
+    #    only works on full file names
+    sensor_str = l1a_dataset.attrs["Logical_source"].split("_")[-1].split("-")[0]
+    gattrs = CDF_MANAGER.get_global_attributes("imap_hi_l1b_de_attrs")
```
same here?

tech3371 reviewed (approved):
Just had minor comments. Otherwise, it looks good to me. Nice refactoring!

greglucas commented on imap_processing/hi/utils.py:
```diff
@@ -18,3 +18,15 @@ class HIAPID(IntEnum):
+        Returns
+        -------
+        str
+            "45" or "90"
```
```suggestion
            "45sensor" or "90sensor"
```

greglucas commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -178,21 +179,27 @@ def create_dataset(de_data_list: list, packet_met_time: list) -> xr.Dataset:
     packet_met_time : list
         List of packet MET time
+    sensor_str : str
+        Sensor head: "45" or "90"
```
```suggestion
        Sensor head: "45sensor" or "90sensor"
```

greglucas commented on imap_processing/hi/l1a/science_direct_event.py:
```diff
@@ -245,86 +252,60 @@ def create_dataset(de_data_list: list, packet_met_time: list) -> xr.Dataset:
+        if var_name == "epoch":
+            continue
+        attrs = cdf_manager.get_variable_attributes(f"hi_de_{var_name}")
+        dtype = attrs.pop("dtype")
```
Nice, I like this and the format style above! Seems like a nice clean way to do this.

I assume that `cdf_manager.get_variable_attributes()` returns a copy of the `attrs` dict each time so you aren't mutating/removing the `dtype` from the original version here?

tmplummer replied:
Good point... I don't see that a copy is being returned. I will copy here for now.

greglucas commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -42,9 +53,67 @@ def hi_l1b(l1a_cdf_path: Path):
+        "quality_flag",
+        "nominal_bin",
+    ]:
+        attrs = CDF_MANAGER.variable_attributes[f"hi_de_{var}"]
```
I think this is where @maxinelasp was concerned with returning mutable instances of things, where the `pop` below will be impacting future lookups. I don't remember where everything stands, but worth hearing from her on this use.
```suggestion
        attrs = CDF_MANAGER.get_variable_attributes(f"hi_de_{var}")
```

tmplummer replied:
I will also use `copy()` here to avoid unwanted mutation.

greglucas commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -42,9 +53,67 @@ def hi_l1b(l1a_cdf_path: Path):
+            dims=["epoch"],
+            attrs=attrs,
         )
+    l1b_dataset = l1a_dataset.assign(new_data_vars)
+    l1b_dataset = l1b_dataset.drop_vars(
+        ["tof_1", "tof_2", "tof_3", "de_tag", "ccsds_met", "meta_event_met"]
+    )
```
I like this style of sticking with xarray and updating your datasets and what is contained at the various levels. 👍

greglucas commented on imap_processing/tests/hi/conftest.py:
```diff
@@ -0,0 +1,92 @@
+            # probability of 0.xx. The `k=1` argument specifies that one
+            # item should be chosen, and `[0]` is used to extract the
+            # single item from the resulting list.
+            tof_1 = random.choices(
```
Is there an issue if these are out of order, `tof_1` is after `tof_2`? I'm wondering if this should actually test any explicit cases like that you know you want to handle/fail properly?

Could you also write this more succinctly as:
`tof_1, tof_2, tof_3 = random.choices([random.randint(0, 1022), 1023], weights=[0.45, 0.55], k=3)`

tmplummer replied:
Good question on the time-of-flight ordering. I suspect there could be issues once the L1B processing is fleshed out a bit more. Coincidence type will also need to agree with the TOF values. For now, this if fine. Hopefully we will have some real test data before this becomes important in testing the processing code.

greglucas commented on imap_processing/tests/hi/conftest.py:
```diff
@@ -0,0 +1,92 @@
+        else:
+            pkt_len = 4 + 4 + 6 * num_directevents + 2 - 1
+        ccsds_data = (
+            ccsds_header_data(770, pkt_len)
```
Could you parameterize/request this from the test fixture so you can test both 45 and 90 apids?
```suggestion
            ccsds_header_data(770, pkt_len)
```

tmplummer committed:
Fix silly double use of double quotes

tmplummer committed:
Changes from PR reviews
Fix docstrings
Copy output from cdf manager to avoid mutation
Parameterize apid in fixture that creates fake DE data

tmplummer committed:
Use typing.Union rather than pipe

greglucas commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -26,12 +33,16 @@ def hi_l1b(l1a_cdf_path: Path):
-    l1a_dataset = load_cdf(l1a_cdf_path)
-    packet_enum = HIAPID(l1a_dataset["pkt_apid"].data[0])
+    logger.info(f"Running Hi L1B processing on file: {l1a_cdf_path.name}")
+    l1a_dataset = load_cdf(l1a_cdf_path, to_datetime=True)
```
After your last PR I think this can be removed now.
```suggestion
    l1a_dataset = load_cdf(l1a_cdf_path)
```

tmplummer replied:
Good catch!

tmplummer committed:
remove unneeded kwarg to load_cdf

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>

tmplummer committed:
Update global attribute YAML to clarify structure
